### PR TITLE
Census of India 2011 - Data Imports

### PR DESCRIPTION
**Pull request by thejeshgn**
- Modular structure to make it easy to import other census datasets in future
- General scripts for importing for census of India Primary Data
- Currently includes scripts for importing three datasets
1. Primary Census Abstract Total (India, States/UTs - District Level)

---
This change is [Reviewable](https://reviewable.io/reviews/datacommonsorg/data/331)

### CODE REVIEW

File: scripts/india_census/common/base.py [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+        return constraints
+
+    def _create_variable(self, data_row, place_of_residence=None):
+        row = copy.deepcopy(data_row)
+        name = self._get_base_name(row)
+        constraints = self._get_base_constraints(row)
+
+        if row["age"] == "YearsUpto6":
+            name = name + "_" + "YearsUpto6"
+            constraints = constraints + "age: dcid:YearsUpto6 \n"
```
Comment (pradh) [inline scripts/india_census/common/base.py:150]: drop trailing space?

Or, could make name and constraints an array that you finally join with '_' and '\n' respectively.

File: scripts/india_census/common/base.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+        else:
+            pass
+
+        if row["workerStatus"] == "Worker":
+            constraints = constraints + "workerStatus: Worker \n"
```
Comment (pradh) [inline scripts/india_census/common/base.py:155]: dcs:Worker

?

(and below)

File: scripts/india_census/common/base.py [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+            if row["workerClassification"] == "MainWorker":
+                name = name + "_" + "MainWorkers"
+                constraints = constraints + "workerClassification: MainWorker \n"
+                if row["workCategory"] != "":
+                    name = name + "_" + row["workCategory"]
+                    constraints = constraints + "workCategory:" + row[
+                        "workCategory"] + " \n"
+
+            elif row["workerClassification"] == "MarginalWorker":
+                name = name + "_" + "MarginalWorkers"
```
Comment (pradh) [inline scripts/india_census/common/base.py:165]: MarginalWorker

File: scripts/india_census/common/base.py [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+            "Level", "Name"
+        ],
+                         axis=1,
+                         inplace=True)
+        #first column is Name of the place
+        #second column is Name of the TRU/placeOfResidence
+        #3-N are the actual values
+        value_columns = list(self.raw_df.columns[1:-1])
+
+        #converting rows in to columns. So the filnal structure will be
```
Comment (pradh) [inline scripts/india_census/common/base.py:114]: final

File: scripts/india_census/common/base.py [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+
+                if row["workPeriod"] == "[Month 3 6]":
+                    name = name + "_" + "Worked3To6Months"
+                    constraints = constraints + "workPeriod:" + row[
+                        "workPeriod"] + " \n"
+            else:
+                name = name + "_" + "Workers"
+
+        elif row["workerStatus"] == "NonWorker":
+            name = name + "_" + "NonWorkers"
```
Comment (pradh) [inline scripts/india_census/common/base.py:185]: NonWorker?

File: scripts/india_census/common/primary_abstract_data_variables.csv [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+P_ILL,Illiterates,Person,measuredValue,count,,,,Illiterate,,,,
+M_ILL,Male Illiterates,Person,measuredValue,count,Male,,,Illiterate,,,,
+F_ILL,Female Illiterates,Person,measuredValue,count,Female,,,Illiterate,,,,
+TOT_WORK_P,Total Workers,Person,measuredValue,count,,,,,Worker,,,
+TOT_WORK_M,Total Male Workers,Person,measuredValue,count,Male,,,,Worker,,,
+TOT_WORK_F,Total Female Workers,Person,measuredValue,count,Female,,,,Worker,,,
+MAINWORK_P,Main Workers,Person,measuredValue,count,,,,,Worker,MainWorker,,
+MAINWORK_M,Male Main Workers,Person,measuredValue,count,Male,,,,Worker,MainWorker,,
+MAINWORK_F,Female Main Workers,Person,measuredValue,count,Female,,,,Worker,MainWorker,,
+MAIN_CL_P,Main Cultivators,Person,measuredValue,count,,,,,Worker,MainWorker,Cultivators,
```
Comment (pradh) [inline scripts/india_census/common/primary_abstract_data_variables.csv:27]: Question whether this should be plural or singular?

@tjann

File: scripts/india_census/common/test_data/test_Data_Expected.mcf [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+typeOf: dcs:StatisticalVariable
+populationType: dcs:Person
+statType: dcs:measuredValue
+measuredProperty: dcs:count
+placeOfResidence: dcs:Rural 
+gender: schema:Female 
+
+
+
+Node: dcid:Count_Person_YearsUpto6
```
Comment (pradh) [inline scripts/india_census/common/test_data/test_Data_Expected.mcf:66]: nit: We could reduce 1 newline :)

File: scripts/india_census/primary_census_abstract_data/readme.md [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+| ....  |
+| ....  |
+
+
+#### Cleaned Data
+As part of the cleaning process, we remove the unwanted columns and then convert the columns into rows. Then add the `StatisticalVariable` column corresponding to census variable and TRU.
+
+Cleaned data [IndiaCensus2011_Primary_Abstract_Data.csv](IndiaCensus2011_Primary_Abstract_Data.csv) will have the following columns.
+
+- census_location_id - Census location id
```
Comment (pradh) [inline scripts/india_census/primary_census_abstract_data/readme.md:40]: To confirm is this census_location_id or indianCensusAreaCode{YYYY} ?

**Review by pradh — COMMENTED** [review]
Great job!  Very cleanly done!

I reviewed:
  - india_census/common/
  - india_census/primary_census_abstract_data/

In the interest of keeping the PR of a reasonable size, could we move the other subdirs to a follow on PR?

File: scripts/india_census/common/base.py [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+        dtype = {
+            'State': str,
+            'District': str,
+            'Subdistt': str,
+            "Town/Village": str,
+            "Ward": str,
+            "EB": str
+        }
+        self.raw_df = pd.read_excel(self.data_file_path, dtype=dtype)
+        self.census_columns = self.raw_df.columns[9:]
```
[the remark this answers is not in the document]
Comment (thejeshgn) [reply scripts/india_census/common/base.py]: Done.

File: scripts/india_census/common/base.py (hunk 1 above)
Comment (thejeshgn) [reply scripts/india_census/common/base.py]: Done.

File: scripts/india_census/common/primary_abstract_data_variables.csv (hunk 6 above)
Comment (thejeshgn) [reply scripts/india_census/common/primary_abstract_data_variables.csv:27]: I have converted into singular to keep it uniform. Let me know.

File: scripts/india_census/common/test_data/test_Data_Expected.mcf (hunk 7 above)
Comment (thejeshgn) [reply scripts/india_census/common/test_data/test_Data_Expected.mcf:54]: 👍🏼

File: scripts/india_census/primary_census_abstract_data/readme.md (hunk 8 above)
Comment (thejeshgn) [reply scripts/india_census/primary_census_abstract_data/readme.md]: We are using  indianCensusAreaCode{YYYY}/indianCensusAreaCode2011 I have left the csv column name as it is.

I have updated the documentation.

File: scripts/india_census/common/base.py (hunk 5 above)
Comment (thejeshgn) [reply scripts/india_census/common/base.py]: Done.

File: scripts/india_census/common/base.py (hunk 3 above)
Comment (thejeshgn) [reply scripts/india_census/common/base.py]: Done.

File: scripts/india_census/common/base.py [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+indianCensusAreaCode{year}: C:IndiaCensus{year}_{dataset_name}->census_location_id
+
+"""
+
+
+class CensusPrimaryAbstractDataLoaderBase:
+
+    def __init__(self, data_file_path, metadata_file_path, mcf_file_path,
+                 tmcf_file_path, csv_file_path, existing_stat_var, census_year,
+                 dataset_name):
```
[the remark this answers is not in the document]
Comment (thejeshgn) [reply scripts/india_census/common/base.py:66]: 👍🏼

**Comment by thejeshgn** [comment]
> In the interest of keeping the PR of a reasonable size, could we move the other subdirs to a follow on PR?

I have removed other datasets for this PR. Once this gets merged. I will submit all of them as independent PRs.

File: scripts/india_census/common/primary_abstract_data_variables.csv (hunk 6 above)
Comment (pradh) [reply scripts/india_census/common/primary_abstract_data_variables.csv:27]: Sorry, I mean things like: Cultivators, AgriculturalLabourers, HouseholdIndustries

**Review by pradh — APPROVED** [review]

File: scripts/india_census/common/base.py [hunk 11]
```diff
@@ -26,12 +26,10 @@
 typeOf: dcs:StatisticalVariable
 populationType: dcs:{populationType}
 statType: dcs:{statType}
-measuredProperty: dcs:{measuredProperty}
-{constraints}
+measuredProperty: dcs:{measuredProperty}{constraints}
```
[the remark this answers is not in the document]
Comment (thejeshgn) [reply scripts/india_census/common/base.py]: 1. I don't add empty constraints to the list anymore. That should solve.
2. I also move space separator in MCF file from top to bottom

**Comment by thejeshgn** [comment]
> Sorry, I mean things like: Cultivators, AgriculturalLabourers, HouseholdIndustries

From our notes. `workCategory will Cultivators or AgriculturalLabourers or HouseholdIndustries or OtherWorkers`

 @tjann @pradh Shall I make it singular?

**Comment by pradh** [comment]
> > Sorry, I mean things like: Cultivators, AgriculturalLabourers, HouseholdIndustries
> 
> From our notes. `workCategory will Cultivators or AgriculturalLabourers or HouseholdIndustries or OtherWorkers`
> 
> @tjann @pradh Shall I make it singular?

Since this is the only pending thing.  I'll go ahead and merge this change now.   And we can resolve it in the next round.

**Pull request merged**